### PR TITLE
Migrates change_url_modal.jsx to be pure and use Redux 

### PR DESCRIPTION
**Pull request by watadarkstar**
#### Summary

This PR migrates `change_url_modal.jsx` to be pure and use Redux. It also cleans up a bunch of code surrounding `new_channel_modal.jsx` as I found most of this flow to be outdated and or broken which I needed to fix first to test `change_url_modal.jsx`.

- Changed `Component` to `PureComponent`
- Replaced old `.bind(this)` syntax with corresponding `() =>` syntax. 
- Moved `PropTypes` to inside class and documented them
- Removed calling `this.ref.urlinput.select()` on input DOM node as `this.ref.urlinput` was `null` inside `componentDidUpdate` and removed `componentDidUpdate` since we don't need it anymore. I'm not even sure why there was `this.ref.urlinput.select()` I assume to select the input but this wasn't working. CC: @jwilander 
- Removed `ReactDOM.findDOMNode` as [its not advised by the linter and will be deprecated eventually](https://github.com/yannickcr/eslint-plugin-react/issues/678) when you can use the `ref` node directly instead
- Changed `doSubmit` to `onSubmit` to be consistent
- Moved `change_url_modal.jsx` to its own folder
- Removed `SHOW_EDIT_URL_THEN_COMPLETE` because this is redundant as `new_channel_modal.jsx`  is already checking to see that the channel name is of length > 2 i.e. https://github.com/mattermost/mattermost-webapp/blob/b28ae6e6fc4e7458171939eb20a1c5cf36d6cae9/components/new_channel_modal/new_channel_modal.jsx#L130-L132
- Fix `enterKeyDown` inside `new_channel_modal.jsx` from blocking `enterKeyDown` in the `change_url_modal.jsx`. I got rid of the `document. addEventListener `  inside `new_channel_modal.jsx` and now only check for key down on the `<input>` in which it was necessary.
- Updated Jest snapshots
- Cleaned up `new_channel_flow` to make use of the constants file

#### Ticket Link
https://github.com/mattermost/mattermost-server/issues/7658

#### Checklist

- [x] Convert to Pure Component
- [x] Convert to use Redux
- [x] Fix failing tests
- [x] Write tests for `ChangeUrlModal`
- [x] Write tests for `NewChannelFlow`
- [x] Ran `make check-style` to check for style errors (required for all pull requests)
- [x] Ran `make test` to ensure unit and component tests passed
- [x] Added or updated unit tests (required for all new features)
- [x] ~~Has server changes (please link)~~
- [x] ~~Has redux changes (please link)~~
- [x] ~~Has UI changes~~
- [x] ~~Includes text changes and localization file ([.../i18n/en.json](https://github.com/mattermost/mattermost-webapp/blob/master/i18n/en.json)) updates~~
- [x] ~~Touches critical sections of the codebase (auth, posting, etc.)~~

### CODE REVIEW

File: components/new_channel_modal/new_channel_modal.jsx [hunk 1]
```diff
@@ -264,7 +260,7 @@ export default class NewChannelModal extends React.PureComponent {
                                 <div className='col-sm-9'>
                                     <input
                                         id={inputPrefixId + 'Name'}
-                                        onChange={this.handleChange}
+                                        onChange={() => this.handleChange()}
```
Comment (saturninoabril) [inline components/new_channel_modal/new_channel_modal.jsx]: Should not pass anonymous function as props. This should be like:
```
onChange={this.handleChange}
```

then `handleChange` should be:
```
handleChange = () => {
    ...
}
```

Please do the same pattern to `onEnterKeyDown` and `handleSubmit`.

File: components/change_url_modal/change_url_modal.jsx [hunk 2]
```diff
@@ -4,21 +4,72 @@
...
-import ReactDOM from 'react-dom';
 import {FormattedMessage} from 'react-intl';
 
 import TeamStore from 'stores/team_store.jsx';
 
 import Constants from 'utils/constants.jsx';
 import * as URL from 'utils/url.jsx';
 
-export default class ChangeUrlModal extends React.Component {
+export default class ChangeUrlModal extends React.PureComponent {
```
Comment (saturninoabril) [inline components/change_url_modal/change_url_modal.jsx]: Please change component name to `ChangeURLModal` as we agreed on using uppercase form to those abbreviations like `URL`, `HTML`, etc.

File: components/change_url_modal/change_url_modal.jsx [hunk 3]
```diff
@@ -4,21 +4,72 @@
...
+    }
+
     constructor(props) {
         super(props);
 
         this.onURLChanged = this.onURLChanged.bind(this);
-        this.doSubmit = this.doSubmit.bind(this);
-        this.doCancel = this.doCancel.bind(this);
+        this.onSubmit = this.onSubmit.bind(this);
+        this.onCancel = this.onCancel.bind(this);
```
Comment (saturninoabril) [inline components/change_url_modal/change_url_modal.jsx]: Please remove bind to these functions.

File: components/change_url_modal/change_url_modal.jsx [hunk 4]
```diff
@@ -107,10 +152,9 @@ export default class ChangeUrlModal extends React.Component {
         return error;
     }
 
-    doSubmit(e) {
+    onSubmit(e) {
```
Comment (saturninoabril) [inline components/change_url_modal/change_url_modal.jsx]: Please change to 
```
onSubmit = (e) => {
    ...
}
```

Same case to `onURLChanged` and `onCancel`

**Comment by saturninoabril** [comment]
Also, you have a failing test. You may run `make test` to check it locally.

File: components/new_channel_modal/new_channel_modal.jsx (hunk 1 above)
Comment (watadarkstar) [reply components/new_channel_modal/new_channel_modal.jsx]: Oh I didn't know you could do that. Cool, I learned something today!

**Comment by watadarkstar** [comment]
@saturninoabril I fixed your requested changes. Thank you! I still need to writes tests for: `ChangeUrlModal` and `NewChannelFlow`.

File: components/change_url_modal/index.js [hunk 5]
```diff
@@ -0,0 +1,17 @@
+// Copyright (c) 2017 Mattermost, Inc. All Rights Reserved.
+// See License.txt for license information.
+
+import {connect} from 'react-redux';
+
+import {getCurrentTeamUrl} from 'mattermost-redux/selectors/entities/teams';
```
Comment (saturninoabril) [inline components/change_url_modal/index.js]: Hmmm, unfortunately `getCurrentTeamUrl` is not working as expected. Refer to similar implementation here https://github.com/mattermost/mattermost-webapp/blob/master/components/get_post_link_modal/index.js

Comment (watadarkstar) [reply components/change_url_modal/index.js]: @saturninoabril is this because the URL says `undefined` at the beginning. I believe its broken even there.

Comment (watadarkstar) [reply components/change_url_modal/index.js]: Oops nvm I see what you mean!

Comment (watadarkstar) [reply components/change_url_modal/index.js]: Fixed in my latest commit

**Comment by watadarkstar** [comment]
@saturninoabril Fixed your last request. Thank you for pointing that out I was wondering why it was not working. ⭐️

File: components/change_url_modal/index.js [hunk 6]
```diff
@@ -0,0 +1,21 @@
...
+
+import {getSiteURL} from 'utils/url';
+
+import ChangeURLModal from './change_url_modal';
+
+function mapStateToProps(state, ownProps) {
+    const currentTeam = getCurrentTeam(state);
+    const currentTeamURL = `${getSiteURL()}/${currentTeam.name}`;
+    return {
+        ...ownProps,
```
Comment (jwilander) [inline components/change_url_modal/index.js]: Turns out we don't actually need to do`...ownProps` here (I know we're doing it everywhere else), it's handled automatically. Mind just removing this line?

Comment (watadarkstar) [reply components/change_url_modal/index.js]: @jwilander I removed this here: https://github.com/mattermost/mattermost-webapp/pull/464/commits/16139e8d7462077e18c1bed97bc9ec13f9465cd8 and also removed `ownProps` as a param.

**Comment by saturninoabril** [comment]
@watadarkstar Looks good to me. Let me know if you still plan to submit test for `NewChannelFlow`. Thanks!

**Comment by watadarkstar** [comment]
@saturninoabril Yup I am writing the tests right now. Will let you know when I am done hopefully by the latest EOD.

**Comment by saturninoabril** [comment]
Cool, thanks!

**Comment by watadarkstar** [comment]
@saturninoabril Wrote tests and did a whole bunch of cleanup in my last commit for `NewChannelFlow` https://github.com/mattermost/mattermost-webapp/pull/464/commits/5c9454bb6566bb1111495f7a97f9c68d7395f338 . Give it a review and this should be good to merge!

**Comment by watadarkstar** [comment]
@saturninoabril the jenkins build is not working its been going on for an hour - could you restart it?

**Pull request merged**